### PR TITLE
[crypto+move] fix authentication key rotation and type-safe Ed25519 and ECDSA in Move

### PR DESCRIPTION
**Pull request by alinush**
### Description

 1. There was a security issue in the authentication key rotation for Ed25519: we were not verifying that the `old_public_key` matched the current authentication key in the account, making the digital-signature based capability verification useless
 2. Re-wrote the Ed25519 and ECDSA modules to be more type-safe

- - -
This change is [Reviewable](https://reviewable.io/reviews/aptos-labs/aptos-core/3109)

### CODE REVIEW

File: aptos-move/framework/aptos-framework/sources/account.move [hunk 1]
```diff
@@ -205,46 +208,66 @@ module aptos_framework::account {
...
-        let new_auth_key = hash::sha3_256(new_public_key);
+        // Derive the authentication key of the new PK
+        vector::push_back(&mut new_pk_bytes, 0);
+        let new_auth_key = hash::sha3_256(new_pk_bytes);
         let new_address = create_address(new_auth_key);
+
+        // Update the originating address map
         table::add(address_map, new_address, addr);
+
+        // Update the account with the new authentication key
```
Comment (davidiw) [inline aptos-move/framework/aptos-framework/sources/account.move:270]: yuck, these comments are too verbose

**Review by davidiw — APPROVED** [review]

File: aptos-move/framework/src/natives/cryptography/ed25519.rs [hunk 2]
```diff
@@ -131,123 +120,31 @@ fn native_ed25519_verify(
...
  **************************************************************************************************/
 #[derive(Debug, Clone)]
 pub struct GasParameters {
-    // Ed25519 signatures
-    pub ed25519_validate_pubkey: Ed25519ValidatePubkeyGasParameters,
-    pub ed25519_verify: Ed25519VerifyGasParameters,
-
-    // ECDSA signatures based on secp256k1 elliptic curves
-    pub secp256k1_ecdsa_recover: Secp256k1ECDSARecoverGasParameters,
+    pub base_cost: InternalGas,
```
Comment (vgao1996) [inline aptos-move/framework/src/natives/cryptography/ed25519.rs:129]: Are you sure you don't need a different base cost for different functions?

**Review by chloeqjz — APPROVED** [review]

File: aptos-move/framework/src/natives/cryptography/ed25519.rs (hunk 2 above)
Comment (alinush) [reply aptos-move/framework/src/natives/cryptography/ed25519.rs:129]: As far as I can tell, all of these functions do mostly cryptographic work, so using the same base cost is reasonable. @davidiw.

Comment (alinush) [reply aptos-move/framework/src/natives/cryptography/ed25519.rs:129]: ...and from discussing with you, Victor.

**Pull request merged**